artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -380,10 +394,7 @@ private void performAck(String nodeID, long messageID, Queue targetQueue, ACKMes
             logger.warn(e.getMessage(), e);
          }
       } else {
-         if (logger.isDebugEnabled()) {
-            logger.debug("Post ack Server " + server + " could not find messageID = " + messageID +
-                            " representing nodeID=" + nodeID);
-         }
+         performAckOnPage(nodeID, messageID, targetQueue, ackMessageOperation);
```

Comment:
this is using the same thread that would been used on the page reading process (the scan is doing that).

in most of the cases the message should been on the queue by the depage process. in case things are out of balance, the scan will read on the paging... using the same thread that would been used on reading from page. it will then issue the pageACK.  I don't think you could get messages orphaned on this case.

Reply:
The scan is done on the same thread that reads the pages, yes...but I dont see where there is any guarantee or coordination that the message hasnt already depaged by that thread before we eventually get round to asking it to do the scan for us, after trying and failing to ack the message on the queue itself from the other IO thread?

To me it seems like the connection thread can try to ack on the queue, fail to do so as it isnt there yet...decide to pop a scan onto the paging thread, only for that thread to successfully depage the message while we are preparing the request, and then be asked to remove it from the paging store and fail at that because its no longer there anymore by the time it runs, but is now on the queue we wont check again.

Reply:
hmmm... let me think about that.. I will address this shortly.

Reply:
#3737 is addressing this